File: backend/src/main/kotlin/com/tnmaster/application/controller/AddressV2Api.kt
package com.tnmaster.application.controller

import cn.dev33.satoken.annotation.SaCheckPermission
import com.tnmaster.dto.address.AddressFullPathView
import com.tnmaster.entities.Address
import com.tnmaster.application.repositories.IAddressRepo
import com.tnmaster.application.service.AddressService
import org.babyfish.jimmer.client.meta.Api
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

/** ## 第2版地址 API */
@Api
@RestController
@RequestMapping("v2/address")
class AddressV2Api(
  private val repo: IAddressRepo,
  private val addressService: AddressService,
) {
  /** ## 获取直接的地址直接子集 */
  @Api
  @GetMapping("direct_children")
  fun getDirectChildrenByCode(@RequestParam code: String): List<Address> {
    return addressService.fetchDirectChildrenByCode(code = code)
  }

  /**
   * ## 批量获取地址全路径
   *
   * @param codes 批量地址代码
   */
  @Api
  @GetMapping("full_paths/codes")
  fun getFullPathsByCodes(@RequestParam codes: List<String>): List<AddressFullPathView> {
    require(codes.size <= 100) { "codes.size must be less than or equal to 100" }
    return repo.findAllFullPathByCodesIn(codes)
  }

  /** ## 查询所有省份 */
  @Api
  @GetMapping("provinces")
  fun getAllProvinces(): List<Address> {
    return addressService.fetchProvince()
  }

  /**
   * ## 初始化省份
   * TODO ROOT 权限
   */
  @Api
  @SaCheckPermission("ROOT")
  @PostMapping("init_provinces")
  fun postInitProvincesAsAdmin() {
    addressService.initProvince()
  }
}

File: frontend/src/__generated/services/AddressV2Api.ts
import type { Executor } from '../'
import type { Dynamic_Address } from '../model/dynamic/'
import type { AddressFullPathView } from '../model/static/'

/**
 * ## 第2版地址 API
 */
export class AddressV2Api {
  constructor(private executor: Executor) {}

  /**
   * ## 查询所有省份
   */
  readonly getAllProvinces: () => Promise<
    Array<Dynamic_Address>
  > = async () => {
    const _uri = '/v2/address/provinces'
    return (await this.executor({ uri: _uri, method: 'GET' })) as Promise<Array<Dynamic_Address>>
  }

  /**
   * ## 获取直接的地址直接子集
   */
  readonly getDirectChildrenByCode: (options: AddressV2ApiOptions['getDirectChildrenByCode']) => Promise<
    Array<Dynamic_Address>
  > = async (options) => {
    let _uri = '/v2/address/direct_children'
    let _separator = !_uri.includes('?') ? '?' : '&'
    let _value: any
    _value = options.code
    _uri += _separator
    _uri += 'code='
    _uri += encodeURIComponent(_value)
    _separator = '&'
    return (await this.executor({ uri: _uri, method: 'GET' })) as Promise<Array<Dynamic_Address>>
  }

  /**
   * ## 批量获取地址全路径
   *
   * @parameter {AddressV2ApiOptions['getFullPathsByCodes']} options
   * - codes 批量地址代码
   */
  readonly getFullPathsByCodes: (options: AddressV2ApiOptions['getFullPathsByCodes']) => Promise<
    Array<AddressFullPathView>
  > = async (options) => {
    let _uri = '/v2/address/full_paths/codes'
    let _separator = !_uri.includes('?') ? '?' : '&'
    let _value: any
    _value = options.codes
    for (const _item of _value) {
      _uri += _separator
      _uri += 'codes='
      _uri += encodeURIComponent(_item)
      _separator = '&'
    }
    return (await this.executor({ uri: _uri, method: 'GET' })) as Promise<Array<AddressFullPathView>>
  }

  /**
   * ## 初始化省份
   * TODO ROOT 权限
   */
  readonly postInitProvincesAsAdmin: () => Promise<
    void
  > = async () => {
    const _uri = '/v2/address/init_provinces'
    return (await this.executor({ uri: _uri, method: 'POST' })) as Promise<void>
  }
}

export interface AddressV2ApiOptions {
  getDirectChildrenByCode: {
    code: string
  }
  getFullPathsByCodes: {
    /**
     * 批量地址代码
     */
    codes: Array<string>
  }
  getAllProvinces: {}
  postInitProvincesAsAdmin: {}
}
